### ISSUE TITLE
taille des noms de bases de données

### ISSUE THREAD

**Issue by asso-infini** [AlternC/AlternC#75]
version : 3.2.10
La taille des noms des bases de données créées depuis le panel ne peut excéder 16 caractères, login et underscore compris. Cette limite n'existait pas en 0.9 et est assez restrictive, quand on a un login de 10 caractères, il n'en reste que 5 pour la base, ça devient un peu juste pour s'y retrouver.

La raison est la suivante, le script _sql_list.php_ l'interdit :
// Max 16 caracters for the database name if we want the mysql user to be automatically created.
Ce n'est donc pas mysql qui l'interdit, c'est le champ _name_ de la table _dbusers_ qui est limité à 16 (historiquement j'imagine, depuis le début d'alternc).

Ne pourrait-on pas augmenter la taille du champ _name_ à au moins 24 ou 32 et modifier le _sql_list.php_ en conséquence ?
Ou bien générer un dbuser en le tronquant à 16 car. maxi (en s'assurant qu'il n'existe pas déjà, quitte à le suffixer d'un numéro) ?

**Comment by vincib** [comment]
MySQL username can only be 16 characters, and 32 from 5.7.8+.
AlternC limits the login name to 14 characters on purpose (see m_admin.php ligne 606)
For me, this is not really a bug, we can't do much on it, unless we are ready to manage a *mess* between DB NAME and DB USERNAME.

> wontfix

**Comment by innoticFR** [comment]
Limiting database name to 16 characters when the default limit is 64 should be reconsidered IMHO.

It's true that since AlternC has evolved by creating one new database user for each new database, we should have a common limit, but, now that MySql has also evolved, a limit of 32 chars would be a good compromise.

One solution would be to use the result of the following command to fix the limit:

    SELECT CHARACTER_MAXIMUM_LENGTH length
      FROM information_schema.columns 
      WHERE TABLE_SCHEMA='mysql' and TABLE_NAME='user' and COLUMN_NAME='User';

This might be executed during `alternc.install`

**Comment by innoticFR** [comment]
I reopen as I don't think this would imply a lot of work

**Comment by vincib** [comment]
MySQL 5.7.8 is **not even** available on SID ... 
and having ONE user for EACH MySQL DB seems like a **very good practice** to me. Do I'd stick to it.

Is it **really** a problem to limit LOGIN names to 14 characters and DB names to 16? I'd say it's not... It's not as if those login are publicly visible anywhere.

**Comment by innoticFR** [comment]
I totally agree with you, but as we will have to implement this one day, the sooner, the better.

Note that the current ticket was about "database" names, not login. Regression are, for users, something very difficult to understand, for instance, when they move their bases from one host with AlternC, to another with a more recent version of AlternC.

Also, MariaDB has a limit of 80 characters (https://mariadb.com/kb/en/mariadb/identifier-names/, starting with MariaDB 5.5.31) for user names and 64 characters for database names. Current versions of MariaDB (10.x) no longer have restrictions for user names, but still have for database names.

============
To conclude, we should expect to have the length of the login name at less equal to the database name.

So we have to : 
 1.  get $sqlserver_login_name_length_max from the request in my previous comment
 2. get $sqlserver_database_name_length_max from a similar request
 3. Set max database length name to max($sqlserver_login_name_length_max, $sqlserver_database_name_length_max)
 4. Set max login length name to $sqlserver_login_name_length_max

**Comment by asso-infini** [comment]
C'est vrai, je n'avais pas précisé, on utilise mariaDB, la limitation de la taille des noms d'utilisateurs est de 80 caractères.
Quand on héberge plusieurs sites sur le même compte (ou plusieurs versions d'un même site), avec plusieurs bases de données, ça peut être compliqué de s'y retrouver avec si peu de caractères, surtout avec un login qui peut prendre 10 ou 12 caractères, hérités d'une migration d'une version précédente. :-/
Ça pourrait être pratique du coup d'avoir un outil de renommage de login (et tout ce qui va avec, répertoire, comptes ftp, comptes mysql etc.)

**Comment by albancrommer** [comment]
As long as MariaDB is not the default databse in AlternC, I think this issue is a no-go, since it could break a lot of installations.

But as it is a limitation and there are solutions, why not think of a workaround ?

Can't one code a plugin / package for alternc containing overrides like the one you propose? i.e. alternc-mariadb?

**Comment by innoticFR** [comment]
I partially fixed this bug. I don't understand the hierarchy in `install/upgrades` as this refers to not existing 3.4.x versions.

We have to add the following line in the migration script

     ALTER IGNORE TABLE db MODIFY db VARCHAR(128), login VARCHAR(128);
     ALTER IGNORE TABLE dbusers MODIFY name VARCHAR(128);

**Comment by innoticFR** [comment]
Note: When I say "partially", it means that the only  remaining thing to do is to add the "ALTER" lines in the migration process (and I am not sure to understand where I should put those lines).

The maximum arbitrary limit is 128 because it is still over known limitations for MariaDB for database names, and for users I didn't want to change to a TEXT field in the mysql DB as such fields lack of indexes.

I choose the easiest solution which is using the variable class, and initializing the values the first time (in config.php) with server values. So for a fresh installation, values will be up to the real possibilities, and the modification will have to be made manually after that.
